File: sdk/luma/state.py
from typing import Any, Dict, List, Optional

from ._http import Http


class StateClient:
    """KV store with JSON values, CAS optimistic locking, TTL, and secondary indexes."""

    def __init__(self, http: Http):
        self._http = http

    def list(self, prefix: Optional[str] = None, limit: int = 100) -> list:
        params: Dict = {"limit": limit}
        if prefix:
            params["prefix"] = prefix
        return self._http.get("/v1/state", params=params)

    async def alist(self, prefix: Optional[str] = None, limit: int = 100) -> list:
        params: Dict = {"limit": limit}
        if prefix:
            params["prefix"] = prefix
        return await self._http.aget("/v1/state", params=params)

    def get(self, key: str) -> dict:
        return self._http.get(f"/v1/state/{key}")

    async def aget(self, key: str) -> dict:
        return await self._http.aget(f"/v1/state/{key}")

    def put(self, key: str, value: Any, ttl_ms: Optional[int] = None,
            if_revision: Optional[int] = None) -> dict:
        """Set a key. Pass if_revision to enable CAS — raises LumaConflictError on mismatch."""
        return self._http.put(f"/v1/state/{key}", _put_body(value, ttl_ms, if_revision))

    async def aput(self, key: str, value: Any, ttl_ms: Optional[int] = None,
                   if_revision: Optional[int] = None) -> dict:
        return await self._http.aput(f"/v1/state/{key}", _put_body(value, ttl_ms, if_revision))

    def delete(self, key: str) -> dict:
        return self._http.delete(f"/v1/state/{key}")

    async def adelete(self, key: str) -> dict:
        return await self._http.adelete(f"/v1/state/{key}")

    def batch_put(self, operations: List[Dict]) -> dict:
        return self._http.post("/v1/state/batch_put", {"operations": operations})

    async def abatch_put(self, operations: List[Dict]) -> dict:
        return await self._http.apost("/v1/state/batch_put", {"operations": operations})

    def create_index(self, field: str) -> dict:
        """Register an in-memory secondary index on a value JSON field. Not persisted — recreate after restart."""
        return self._http.post("/v1/state/indexes", {"field": field})

    async def acreate_index(self, field: str) -> dict:
        return await self._http.apost("/v1/state/indexes", {"field": field})

    def query_index(self, field: str, value: str, limit: int = 100) -> list:
        """O(1) lookup by indexed field value. Requires a prior create_index call for this field."""
        return self._http.get(f"/v1/state/index/{field}/{value}", params={"limit": limit})

    async def aquery_index(self, field: str, value: str, limit: int = 100) -> list:
        return await self._http.aget(f"/v1/state/index/{field}/{value}", params={"limit": limit})


def _put_body(value: Any, ttl_ms: Optional[int], if_revision: Optional[int]) -> Dict:
    body: Dict = {"value": value}
    if ttl_ms is not None:
        body["ttl_ms"] = ttl_ms
    if if_revision is not None:
        body["if_revision"] = if_revision
    return body

File: sdk/typescript/src/state.ts
import { HttpClient } from "./http.js";
import type {
  BatchPutOperation,
  JsonValue,
  StateEntry,
  StatePutOptions,
} from "./types.js";

export class StateClient {
  constructor(private readonly http: HttpClient) {}

  list(
    prefix?: string,
    limit = 100,
  ): Promise<StateEntry[]> {
    const params: Record<string, string | number | boolean | undefined> = { limit };
    if (prefix !== undefined) params["prefix"] = prefix;
    return this.http.get("/v1/state", params);
  }

  get(key: string): Promise<StateEntry> {
    return this.http.get(`/v1/state/${encodeKey(key)}`);
  }

  put(
    key: string,
    value: JsonValue,
    opts: StatePutOptions = {},
  ): Promise<StateEntry> {
    const body: Record<string, unknown> = { value };
    if (opts.ttl_ms !== undefined) body["ttl_ms"] = opts.ttl_ms;
    if (opts.if_revision !== undefined) body["if_revision"] = opts.if_revision;
    return this.http.put(`/v1/state/${encodeKey(key)}`, body);
  }

  delete(key: string): Promise<void> {
    return this.http.delete(`/v1/state/${encodeKey(key)}`);
  }

  batchPut(
    operations: BatchPutOperation[],
  ): Promise<Record<string, unknown>> {
    return this.http.post("/v1/state/batch_put", { operations });
  }

  createIndex(field: string): Promise<Record<string, unknown>> {
    return this.http.post("/v1/state/indexes", { field });
  }

  queryIndex(
    field: string,
    value: string,
    limit = 100,
  ): Promise<StateEntry[]> {
    return this.http.get(`/v1/state/index/${encodeURIComponent(field)}/${encodeURIComponent(value)}`, { limit });
  }
}

function encodeKey(key: string): string {
  // Keys may contain slashes intentionally; only encode unsafe chars.
  return key.split("/").map(encodeURIComponent).join("/");
}
